src/hooks/useUpdateSearchParams.ts

```diff
@@ -0,0 +1,59 @@
+import { useSearchParams } from 'react-router-dom';
+
+type UpdateSearchParamsOptions = {
+  method?: 'push' | 'replace';
+};
+
+type UpdateSearchParams = (
+  key: string,
+  value: string | null,
+  options?: UpdateSearchParamsOptions,
+) => void;
+
+/**
+ * This hook provides an `updateSearchParams` function that can update the
+ * search paramters _without_ creating a rerender on the router level.
+ *
+ * NOTE: It's important to handle the search param you're adding in the react
+ * router loader to make sure that loading a URL from browser or refreshing the
+ * page in browser handles it properly.
+ */
+const useUpdateSearchParams = (): UpdateSearchParams => {
+  // useSearchParams hook provides the current search parameters
+  // and a method to update them, which would trigger a re-render.
+  const [searchParams] = useSearchParams();
+
+  // This function updates the search params without causing a re-render
+  // and provides flexibility to use either pushState or replaceState.
+  const updateSearchParams: UpdateSearchParams = (
+    key,
+    value,
```

Comment:
In terms of API I'd take an object so that we can set several params in one call.